File: 0230-kth-smallest-element-in-a-bst/0230-kth-smallest-element-in-a-bst.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int kthSmallest(TreeNode* root, int k) {
        vector<int> out;

        DFS(root, out);

        return out[k-1];
    }

    void DFS(TreeNode* root, vector<int> &out)
    {
        if(!root) return;

        DFS(root->left, out);
        out.push_back(root->val);
        DFS(root->right, out);

        return;
    }
};

File: 0230-kth-smallest-element-in-a-bst/0230-kth-smallest-element-in-a-bst.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

public class Solution {
    public int count = 0, target = 0;
    public int KthSmallest(TreeNode root, int k) {
        DFS(root, k);
        return target;
    }

    public void DFS(TreeNode root, int k)
    {
        if (root == null) return;
        
        DFS(root.left, k);
        count++;  
        if (count == k)
        {
            target = root.val;
            return;
        }
        DFS(root.right, k);         
       
    }
}
